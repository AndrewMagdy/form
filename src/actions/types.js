export const SUBMIT_SUCCESS = "SUBMIT_SUCCESS";
export const SUBMIT_FAIL = "SUBMIT_FAIL";

export const SHOW_DIALOG = "SHOW_DIALOG";
export const HIDE_DIALOG = "HIDE_DIALOG";

export const VALIDATE_IBAN_REQUEST = "VALIDATE_IBAN_REQUEST";
export const VALIDATE_IBAN_SUCCESS = "VALIDATE_SUCCESS";
export const VALIDATE_IBAN_FAIL = "VALIDATE_FAIL";
