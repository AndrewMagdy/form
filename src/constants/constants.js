export const registrationFormId = "registrationForm";
export const delayDur = 3000;
