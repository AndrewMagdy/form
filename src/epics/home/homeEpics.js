import { of } from "rxjs";
import { ofType } from "redux-observable";
import { ajax } from "rxjs/ajax";
import {
  switchMap,
  map,
  catchError,
  retry,
  delay,
  concat,
  exhaustMap
} from "rxjs/operators";
import {
  showDialog,
  hideDialog,
  validateIbanSuccess,
  validateIbanFail
} from "../../actions/home/homeActions";
import { delayDur } from "../../constants/constants";
import * as actionsTypes from "../../actions/types";

export const submitEpic = action$ =>
  action$.pipe(
    ofType(actionsTypes.SUBMIT_SUCCESS),
    exhaustMap(action =>
      of(showDialog()).pipe(concat(of(hideDialog()).pipe(delay(delayDur))))
    )
  );

export const validateEpic = action$ =>
  action$.pipe(
    ofType(actionsTypes.VALIDATE_IBAN_REQUEST),
    switchMap(action => {
      let { resolve, reject } = action.meta || {};
      return ajax
        .post(process.env.REACT_APP_API_ENDPOINT, action.payload, {
          "Content-Type": "application/json"
        })
        .pipe(
          map(response => {
            if (response.response.valid) {
              if (resolve) {
                resolve();
                return validateIbanSuccess();
              }
            } else {
              if (reject) {
                reject(response.response);
                return validateIbanFail();
              }
            }
          }),
          retry(2),
          catchError(error => {
            if (reject) {
              reject(error);
            }
            return of({
              type: validateIbanFail()
            });
          })
        );
    })
  );
