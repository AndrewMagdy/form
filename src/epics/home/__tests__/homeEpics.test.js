import { ActionsObservable } from "redux-observable";
import * as actionsTypes from "../../../actions/types";
import { submitEpic, validateEpic } from "../homeEpics";
import { ajax } from "rxjs/ajax";
import { of } from "rxjs";
import { TestScheduler } from "rxjs/testing";
import { delayDur } from "../../../constants/constants";

jest.mock("rxjs/ajax");

describe("HomeEpics", () => {
  jest.useFakeTimers();
  jest.runAllTimers();

  test("should submit successfly", () => {
    const testScheduler = new TestScheduler((actual, expected) => {
      expect(actual).toEqual(expected);
    });

    testScheduler.run(({ hot, expectObservable }) => {
      const action$ = hot("-a", {
        a: { type: actionsTypes.SUBMIT_SUCCESS }
      });
      const state$ = null;
      const output$ = submitEpic(action$, state$);

      /*
        https://github.com/ReactiveX/rxjs/blob/master/docs_app/content/guide/testing/marble-testing.md
        NOTE: You may have to subtract 1 millisecond from the time you want to progress because 
        the alphanumeric marbles (representing an actual emitted value) advance time 1 virtual frame
         themselves already, after they emit. This can be very unintuitive and frustrating, 
         but for now it is indeed correct.
         Thats 2999ms instead of 3000ms

        */
      expectObservable(output$).toBe(`-a ${delayDur - 1}ms b`, {
        a: {
          type: actionsTypes.SHOW_DIALOG
        },
        b: {
          type: actionsTypes.HIDE_DIALOG
        }
      });
    });
  });

  test("Should validate successfly", done => {
    const action$ = ActionsObservable.of({
      type: actionsTypes.VALIDATE_IBAN_REQUEST,
      meta: { resolve: () => {}, reject: () => {} }
    });

    ajax.post.mockImplementation(() => {
      return of({ response: { valid: true } });
    });

    validateEpic(action$).subscribe(actual => {
      expect(actual).toMatchSnapshot();
      done();
    });
  });

  test("Should fail validation", done => {
    const action$ = ActionsObservable.of({
      type: actionsTypes.VALIDATE_IBAN_REQUEST,
      meta: { resolve: () => {}, reject: () => {} }
    });

    ajax.post.mockImplementation((url, headers) => {
      return of({ response: { valid: false } });
    });

    validateEpic(action$).subscribe(actual => {
      expect(actual).toMatchSnapshot();
      done();
    });
  });
});
