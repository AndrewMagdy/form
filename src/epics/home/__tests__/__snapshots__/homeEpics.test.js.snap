// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeEpics Should fail validation 1`] = `
Object {
  "type": "VALIDATE_FAIL",
}
`;

exports[`HomeEpics Should validate successfly 1`] = `
Object {
  "type": "VALIDATE_SUCCESS",
}
`;
